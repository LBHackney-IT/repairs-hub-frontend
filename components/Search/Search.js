import { useState, useEffect, useRef } from 'react'
import { useRouter } from 'next/router'
import PropertiesTable from 'components/Properties/PropertiesTable'
import Button from 'components/Form/Button/Button'
import Spinner from 'components/Spinner/Spinner'
import ErrorMessage from 'components/Errors/ErrorMessage/ErrorMessage'
import { getProperties } from 'utils/api/properties'

const Search = ({ query }) => {
  const [searchParams, setSearchParams] = useState('')
  const [properties, setProperties] = useState([])
  const [loading, setLoading] = useState(false)
  const [searching, setSearching] = useState(true)
  const [searchQuery, setSearchQuery] = useState('')
  const [error, setError] = useState()
  const router = useRouter()

  useEffect(() => {
    if(query) {
      setSearchParams(decodeURI(query.q))
      searchForProperties(query.q)
    }
  }, [])

  async function searchForProperties(newSearchQuery) {
    setSearching(true)
    setLoading(true)
    setSearchQuery(newSearchQuery)
    setError(null)

    try {
      const data = await getProperties(newSearchQuery)
      setProperties(data)
    } catch (e) {
      setProperties(null)
      console.log('An error has occured:', e.response)
      setError(`Oops an error occurred with error status: ${e.response?.status}`)
    }

    setSearching(false)
    setLoading(false)
  }

  const handleSubmit = (e) => {
    e.preventDefault()

    setSearchQuery(searchParams)

    updateSearchQuery(searchParams)
    query && searchForProperties(searchParams)
  }

  const updateSearchQuery = (newSearchQuery) => {
    router.push(
      {
        pathname: '/properties/search',
        query: {
          q: encodeURI(newSearchQuery)
        }
      }
    )
  }

  const renderSearchResults = () => {
    if(properties?.length > 0) {
      return <PropertiesTable properties={properties} query={searchQuery} />
    }

    if (!error) {
      return (
        <>
          <div>
            <hr className="govuk-section-break govuk-section-break--l govuk-section-break--visible" />
            <p className="govuk-heading-s">We found 0 matching results for: {decodeURI(searchQuery)}</p>
          </div>
        </>
      )
    }
  }

  return (
    <div>
      <section className="section">
        <h1 className="govuk-heading-m">Find property</h1>

        <div className="govuk-form-group">
          <form>
            <label
              className="govuk-label">
              <p className="govuk-body-s govuk-!-margin-bottom-0">Search by postcode or address</p>
            </label>
            <input
              type="text"
              className="govuk-input govuk-input--width-10 focus-colour govuk-!-margin-bottom-5"
              value={searchParams}
              onChange={event => setSearchParams(event.target.value)}
            />
            <Button type="submit" label="Search" onClick={handleSubmit} />
          </form>
        </div>
      </section>

      {loading ? (
        <Spinner />
      ) : (
        <>
          {!searching && renderSearchResults()}
          {error && <ErrorMessage label={error} />}
        </>
      )}
    </div>
  )
}

export default Search
