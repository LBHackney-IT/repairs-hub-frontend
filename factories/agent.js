import { Factory } from 'fishery'

const agentUserFactory = Factory.define(() => ({
  name: 'An Agent',
  email: 'an.agent@hackney.gov.uk',
  roles: ['agent'],
  hasRole: true,
  hasAgentPermissions: true,
  hasContractorPermissions: false,
  hasContractManagerPermissions: false,
  hasAuthorisationManagerPermissions: false,
  hasOperativePermissions: false,
  hasAnyPermissions: true,
}))

export const agent = agentUserFactory.build()
