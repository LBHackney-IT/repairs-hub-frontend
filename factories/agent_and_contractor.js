import { Factory } from 'fishery'

export const agentAndContractorUserFactory = Factory.define(() => ({
  name: 'An Operative',
  email: 'an.operative@hackney.gov.uk',
  roles: ['agent', 'contractor'],
}))

export const agentAndContractor = agentAndContractorUserFactory.build()
