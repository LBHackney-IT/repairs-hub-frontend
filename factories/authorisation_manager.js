import { Factory } from 'fishery'

export const authorisationManagerUserFactory = Factory.define(() => ({
  name: 'An Authorisation Manager',
  email: 'a.authorisation_manager@hackney.gov.uk',
  roles: ['authorisation_manager'],
}))

export const authorisationManager = authorisationManagerUserFactory.build()
