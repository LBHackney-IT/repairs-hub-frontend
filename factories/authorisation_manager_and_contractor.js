import { Factory } from 'fishery'

const authorisationManagerAndContractorUserFactory = Factory.define(() => ({
  name: 'A DLO Supervisor',
  email: 'a.dlo_supervisor@hackney.gov.uk',
  roles: ['dlo_supervisor'],
  hasRole: true,
  hasAgentPermissions: false,
  hasContractorPermissions: true,
  hasContractManagerPermissions: false,
  hasAuthorisationManagerPermissions: true,
  hasOperativePermissions: false,
  hasAnyPermissions: true,
}))

export const authorisationManagerAndContractor = authorisationManagerAndContractorUserFactory.build()
