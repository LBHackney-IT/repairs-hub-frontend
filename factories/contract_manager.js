import { Factory } from 'fishery'

const contractManagerUserFactory = Factory.define(() => ({
  name: 'A Contract Manager',
  email: 'a.contract_manager@hackney.gov.uk',
  roles: ['contract_manager'],
  hasRole: true,
  hasAgentPermissions: false,
  hasContractorPermissions: false,
  hasContractManagerPermissions: true,
  hasAuthorisationManagerPermissions: false,
  hasOperativePermissions: false,
  hasAnyPermissions: true,
}))

export const contractManager = contractManagerUserFactory.build()
