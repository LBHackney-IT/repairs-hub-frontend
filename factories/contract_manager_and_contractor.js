import { Factory } from 'fishery'

const contractManagerAndContractorUserFactory = Factory.define(() => ({
  name: 'A DLO Contract Manager',
  email: 'a.dlo_contract_manager@hackney.gov.uk',
  roles: ['dlo_contract_manager'],
  hasRole: true,
  hasAgentPermissions: false,
  hasContractorPermissions: true,
  hasContractManagerPermissions: true,
  hasAuthorisationManagerPermissions: false,
  hasOperativePermissions: false,
  hasAnyPermissions: true,
}))

export const contractManagerAndContractor = contractManagerAndContractorUserFactory.build()
