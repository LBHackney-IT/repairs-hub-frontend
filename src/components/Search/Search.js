import { useState, useEffect, useContext } from 'react'
import { useRouter } from 'next/router'
import UserContext from '../UserContext/UserContext'
import PropertiesTable from '../Properties/PropertiesTable'
import { PrimarySubmitButton } from '../Form'
import Spinner from '../Spinner/Spinner'
import ErrorMessage from '../Errors/ErrorMessage/ErrorMessage'
import { getProperties } from '../../utils/frontend-api-client/properties'

const Search = ({ query }) => {
  const { user } = useContext(UserContext)
  const isContractor = user && user.hasContractorPermissions
  const [searchQuery, setSearchQuery] = useState('')
  const [properties, setProperties] = useState([])
  const [loading, setLoading] = useState(false)
  const [error, setError] = useState()
  const router = useRouter()
  const workOrderReferenceRegex = /[0-9]{8}/g

  const searchHeadingText = isContractor
    ? 'Find repair job'
    : 'Find repair job or property'
  const searchLabelText = isContractor
    ? 'Search by work order reference'
    : 'Search by work order reference, postcode or address'

  useEffect(() => {
    if (query) {
      if (workOrderReferenceRegex.test(query) || isContractor) {
        workOrderUrl(decodeURI(query.q))
      } else {
        setSearchQuery(decodeURI(query.q))
        searchForProperties(query.q)
      }
    }
  }, [])

  const searchForProperties = async (searchQuery) => {
    setLoading(true)
    setError(null)

    try {
      const data = await getProperties(searchQuery)
      setProperties(data)
    } catch (e) {
      setProperties(null)
      console.error('An error has occured:', e.response)
      setError(
        `Oops an error occurred with error status: ${e.response?.status} with message: ${e.response?.data?.message}`
      )
    }

    setLoading(false)
  }

  const handleSubmit = (e) => {
    e.preventDefault()

    if (workOrderReferenceRegex.test(searchQuery) || isContractor) {
      workOrderUrl(searchQuery)
    } else {
      propertiesURL(searchQuery)
      searchQuery && searchForProperties(searchQuery)
    }
  }

  const propertiesURL = (searchQuery) => {
    router.push({
      pathname: '/search',
      query: {
        q: encodeURI(searchQuery),
      },
    })
  }

  const workOrderUrl = (searchQuery) => {
    router.push(`/work-orders/${searchQuery}`)
  }

  return (
    <div>
      <section className="section">
        <h1 className="govuk-heading-m">{searchHeadingText}</h1>

        <div className="govuk-form-group">
          <form>
            <label className="govuk-label">
              <p className="govuk-body-s govuk-!-margin-bottom-0">
                {searchLabelText}
              </p>
              <input
                type="text"
                className="govuk-input govuk-input--width-10 focus-colour govuk-!-margin-bottom-5"
                value={searchQuery}
                onChange={(event) => setSearchQuery(event.target.value)}
              />
            </label>
            <PrimarySubmitButton label="Search" onClick={handleSubmit} />
          </form>
        </div>
      </section>

      {loading ? (
        <Spinner />
      ) : (
        <>
          {properties?.length > 0 && (
            <PropertiesTable properties={properties} query={searchQuery} />
          )}
          {error && <ErrorMessage label={error} />}
        </>
      )}
    </div>
  )
}

export default Search
