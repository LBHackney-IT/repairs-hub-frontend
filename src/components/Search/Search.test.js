import { render, screen } from '@testing-library/react'
import UserContext from '../UserContext'
import Search from './Search'
import { agent } from 'factories/agent'
import { contractManager } from 'factories/contract_manager'
import { authorisationManager } from 'factories/authorisation_manager'

jest.mock('next/router', () => ({
  useRouter: jest.fn(),
}))

describe('Search component', () => {
  ;[agent, contractManager, authorisationManager].forEach((role) => {
    describe(`when logged in as ${role}`, () => {
      it('contains text to show that all search types are possible', () => {
        render(
          <UserContext.Provider value={{ user: role }}>
            <Search />
          </UserContext.Provider>
        )

        screen.getByRole('heading', {
          name: 'Find repair work order or property',
        })

        screen.getByLabelText(
          'Search by work order reference, postcode or address'
        )
      })
    })
  })

  it('renders a search form', () => {
    const { asFragment } = render(
      <UserContext.Provider value={{ user: agent }}>
        <Search />
      </UserContext.Provider>
    )

    expect(asFragment()).toMatchSnapshot()
  })
})
