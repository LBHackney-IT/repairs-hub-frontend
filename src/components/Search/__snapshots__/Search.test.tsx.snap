// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search component renders a search form 1`] = `
<DocumentFragment>
  <div>
    <section
      class="section"
    >
      <h1
        class="lbh-heading-h1"
      >
        Find repair work order or property
      </h1>
      <div
        class="govuk-form-group lbh-form-group"
      >
        <form>
          <label
            class="govuk-label lbh-label"
            for="input-search"
          >
            Search by work order reference, postcode or address
          </label>
          <input
            class="govuk-input lbh-input govuk-input--width-10"
            id="input-search"
            name="search-name"
            type="text"
            value=""
          />
          <div
            class="govuk-form-group lbh-form-group"
          >
            <button
              class="govuk-button lbh-button"
              data-module="govuk-button"
              type="submit"
            >
              Search
            </button>
          </div>
        </form>
      </div>
    </section>
  </div>
</DocumentFragment>
`;
